File: ruby/ast_renderer.rb
require_relative 'ast'

# Renders AST nodes as ASCII tree structures
class ASTRenderer
  def render(node)
    lines = render_node(node, "", true)
    lines.join("\n")
  end

  private

  def render_node(node, prefix, is_last)
    lines = []
    
    # Determine the connector and continuation prefix
    connector = is_last ? "└── " : "├── "
    continuation = is_last ? "    " : "│   "
    
    # Add the current node
    lines << prefix + connector + node_label(node)
    
    # Add children recursively
    children = get_children(node)
    children.each_with_index do |child, index|
      is_last_child = (index == children.length - 1)
      child_lines = render_node(child, prefix + continuation, is_last_child)
      lines.concat(child_lines)
    end
    
    lines
  end

  def node_label(node)
    case node
    when AST::IntegerNode
      "Integer: #{node.value}"
    when AST::BinaryOpNode
      "BinaryOp: #{node.operator}"
    when AST::UnaryOpNode
      "UnaryOp: #{node.operator}"
    when AST::RenderNode
      "Render"
    else
      node.class.name
    end
  end

  def get_children(node)
    case node
    when AST::IntegerNode
      []
    when AST::BinaryOpNode
      [node.left, node.right]
    when AST::UnaryOpNode
      [node.operand]
    when AST::RenderNode
      [node.expression]
    else
      []
    end
  end
end

File: node/astRenderer.js
const { TokenType } = require('./lexer');

class ASTRenderer {
  render(node) {
    const lines = this.buildTree(node);
    return lines.join('\n');
  }

  buildTree(node, prefix = '', isLast = true) {
    const lines = [];
    const connector = isLast ? '└── ' : '├── ';
    const nodeLabel = this.getNodeLabel(node);
    
    lines.push(prefix + connector + nodeLabel);
    
    const childPrefix = prefix + (isLast ? '    ' : '│   ');
    const children = this.getChildren(node);
    
    children.forEach((child, index) => {
      const isLastChild = index === children.length - 1;
      lines.push(...this.buildTree(child, childPrefix, isLastChild));
    });
    
    return lines;
  }

  getNodeLabel(node) {
    switch (node.type) {
      case 'NUMBER':
        return `NUMBER(${node.value})`;
      case 'BINARY_OP':
        return `BINARY_OP(${this.getOperatorSymbol(node.operator)})`;
      case 'UNARY_OP':
        return `UNARY_OP(${this.getOperatorSymbol(node.operator)})`;
      case 'RENDER':
        return 'RENDER';
      default:
        return `UNKNOWN(${node.type})`;
    }
  }

  getOperatorSymbol(operator) {
    switch (operator) {
      case TokenType.PLUS:
        return '+';
      case TokenType.MINUS:
        return '-';
      case TokenType.MULTIPLY:
        return '*';
      case TokenType.DIVIDE:
        return '/';
      case TokenType.MODULO:
        return '%';
      case TokenType.EXPONENT:
        return '^';
      case TokenType.FACTORIAL:
        return '!';
      default:
        return operator;
    }
  }

  getChildren(node) {
    switch (node.type) {
      case 'NUMBER':
        return [];
      case 'BINARY_OP':
        return [node.left, node.right];
      case 'UNARY_OP':
        return [node.operand];
      case 'RENDER':
        return [node.expression];
      default:
        return [];
    }
  }
}

module.exports = { ASTRenderer };